internal/controller/merge.go

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controller
+
+import (
+	"fmt"
+
+	"github.com/metallb/frrk8s/internal/frr"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Merges two router configs corresponding to the same ASN+VRF.
+func mergeRouterConfigs(r, toMerge *frr.RouterConfig) (*frr.RouterConfig, error) {
+	err := routersAreCompatible(r, toMerge)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.RouterID == "" {
+		r.RouterID = toMerge.RouterID
+	}
+
+	v4Prefixes := sets.New(append(r.IPV4Prefixes, toMerge.IPV4Prefixes...)...)
+	v6Prefixes := sets.New(append(r.IPV6Prefixes, toMerge.IPV6Prefixes...)...)
+
+	mergedNeighbors, err := mergeNeighbors(r.Neighbors, toMerge.Neighbors)
+	if err != nil {
+		return nil, err
+	}
+
+	r.IPV4Prefixes = sets.List(v4Prefixes)
+	r.IPV6Prefixes = sets.List(v6Prefixes)
+	r.Neighbors = mergedNeighbors
+
+	return r, nil
+}
+
+// Verifies that two routers are compatible for merging.
+func routersAreCompatible(r, toMerge *frr.RouterConfig) error {
+	if r.MyASN != toMerge.MyASN {
+		return fmt.Errorf("different asns (%d != %d) specified for same vrf: %s", r.MyASN, toMerge.MyASN, r.VRF)
+	}
+
+	bothRouterIDsNonEmpty := r.RouterID != "" && toMerge.RouterID != ""
+	routerIDsDifferent := r.RouterID != toMerge.RouterID
+	if bothRouterIDsNonEmpty && routerIDsDifferent {
+		return fmt.Errorf("different router ids (%s != %s) specified for same vrf: %s", r.RouterID, toMerge.RouterID, r.VRF)
+	}
+
+	return nil
+}
+
+// Merges two neighbors slices corresponding to the same router.
+func mergeNeighbors(curr, toMerge []*frr.NeighborConfig) ([]*frr.NeighborConfig, error) {
+	all := curr
+	all = append(all, toMerge...)
+	if len(all) == 0 {
+		return []*frr.NeighborConfig{}, nil
+	}
+
+	mergedNeighbors := map[string]*frr.NeighborConfig{}
+
+	for _, n := range all {
+		curr, found := mergedNeighbors[n.Name]
```

Comment:
it might have been wrong to use the name in the first place, as we won't catch different ASNs for the same neighbor address - switched the key to n.Addr